### PR TITLE
Fix server components externals on SSR layer

### PR DESCRIPTION
**Pull request by huozhi**
### What

Fix the externals resolving for server rendering layer for app router. For SSR requests, if it's next externals, we resolved and return early, if we didn't resolve, keep going through the following externals resolving

### Why

Previously on app router's SSR bundling layer, we didn't go through the following requests when seeing an server external package, it will keep bundling even it's in server components external packages.

A bug found in #61983 

Closes NEXT-2473

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 2092cb9adeeb21681e2eda4abcc8bbc91c1367f3

`pnpm test-dev test/development/basic/barrel-optimization/barrel-optimization.test.ts`
- Skipped in Turbopack > optimizePackageImports > app - should render the icons correctly without creating all the modules
- Skipped in Turbopack > optimizePackageImports > should support MUI
- Skipped in Turbopack > optimizePackageImports > should support "use client" directive in barrel file

<details>
<summary>Expand output</summary>

  ● Skipped in Turbopack › optimizePackageImports › app - should render the icons correctly without creating all the modules

    expect(received).toContain(expected) // indexOf

    Expected substring: "<svg xmlns=\"http://www.w3.org/2000/svg\""
    Received string:    "<!DOCTYPE html><html id=\"__next_error__\"><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"/><link rel=\"preload\" as=\"script\" fetchPriority=\"low\" href=\"/_next/static/chunks/webpack.js\"/><script src=\"/_next/static/chunks/main-app.js\" async=\"\"></script><script src=\"/_next/static/chunks/app-pages-internals.js\" async=\"\"></script><meta name=\"robots\" content=\"noindex\"/><meta name=\"next-error\" content=\"not-found\"/><title>Next.js</title><meta name=\"description\" content=\"Generated by Next.js\"/><script src=\"/_next/static/chunks/polyfills.js\" noModule=\"\"></script></head><body><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script>(self.__next_f=self.__next_f||[]).push([0]);self.__next_f.push([2,null])</script><script>self.__next_f.push([1,\"0:D{\\\"name\\\":\\\"rD\\\"}\\n\"])</script><script>self.__next_f.push([1,\"1:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/app-router.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n7:I[\\\"(app-pages-browser)/./components/slow-component.js\\\",[\\\"app/page\\\",\\\"static/chunks/app/page.js\\\"],\\\"Comp\\\"]\\n9:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca6\"])</script><script>self.__next_f.push([1,\"4815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/layout-router.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\na:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/render-from-template-context.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\nd:I[\\\"(app-pages-browser)/./\"])</script><script>self.__next_f.push([1,\"node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/error-boundary.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n2:D{\\\"name\\\":\\\"\\\"}\\n3:D{\\\"name\\\":\\\"Page\\\"}\\n4:D{\\\"name\\\":\\\"AcademicCapIcon\\\"}\\n\"])</script><script>self.__next_f.push([1,\"4:[\\\"$\\\",\\\"svg\\\",null,{\\\"xmlns\\\":\\\"http://www.w3.org/2000/svg\\\",\\\"viewBox\\\":\\\"0 0 20 20\\\",\\\"fill\\\":\\\"currentColor\\\",\\\"aria-hidden\\\":\\\"true\\\",\\\"aria-labelledby\\\":\\\"$undefined\\\",\\\"children\\\":[null,[\\\"$\\\",\\\"path\\\",null,{\\\"fillRule\\\":\\\"evenodd\\\",\\\"d\\\":\\\"M9.664 1.319a.75.75 0 01.672 0 41.059 41.059 0 018.198 5.424.75.75 0 01-.254 1.285 31.372 31.372 0 00-7.86 3.83.75.75 0 01-.84 0 31.508 31.508 0 00-2.08-1.287V9.394c0-.244.116-.463.302-.592a35.504 35.504 0 013.305-2.033.75.75 0 00-.714-1.319 37 37 0 00-3.446 2.12A2.216 2.216 0 006 9.393v.38a31.293 31.293 0 00-4.28-1.746.75.75 0 01-.254-1.285 41.059 41.059 0 018.198-5.424zM6 11.459a29.848 29.848 0 00-2.455-1.158 41.029 41.029 0 00-.39 3.114.75.75 0 00.419.74c.528.256 1.046.53 1.554.82-.21.324-.455.63-.739.914a.75.75 0 101.06 1.06c.37-.369.69-.77.96-1.193a26.61 26.61 0 013.095 2.348.75.75 0 00.992 0 26.547 26.547 0 015.93-3.95.75.75 0 00.42-.739 41.053 41.053 0 00-.39-3.114 29.925 29.925 0 00-5.199 2.801 2.25 2.25 0 01-2.514 0c-.41-.275-.826-.541-1.25-.797a6.985 6.985 0 01-1.084 3.45 26.503 26.503 0 00-1.281-.78A5.487 5.487 0 006 12v-.54z\\\",\\\"clipRule\\\":\\\"evenodd\\\"}]]}]\\n\"])</script><script>self.__next_f.push([1,\"5:D{\\\"name\\\":\\\"AcademicCapIcon\\\"}\\n\"])</script><script>self.__next_f.push([1,\"5:[\\\"$\\\",\\\"svg\\\",null,{\\\"xmlns\\\":\\\"http://www.w3.org/2000/svg\\\",\\\"fill\\\":\\\"none\\\",\\\"viewBox\\\":\\\"0 0 24 24\\\",\\\"strokeWidth\\\":1.5,\\\"stroke\\\":\\\"currentColor\\\",\\\"aria-hidden\\\":\\\"true\\\",\\\"aria-labelledby\\\":\\\"$undefined\\\",\\\"children\\\":[null,[\\\"$\\\",\\\"path\\\",null,{\\\"strokeLinecap\\\":\\\"round\\\",\\\"strokeLinejoin\\\":\\\"round\\\",\\\"d\\\":\\\"M4.26 10.147a60.436 60.436 0 00-.491 6.347A48.627 48.627 0 0112 20.904a48.627 48.627 0 018.232-4.41 60.46 60.46 0 00-.491-6.347m-15.482 0a50.57 50.57 0 00-2.658-.813A59.905 59.905 0 0112 3.493a59.902 59.902 0 0110.399 5.84c-.896.248-1.783.52-2.658.814m-15.482 0A50.697 50.697 0 0112 13.489a50.702 50.702 0 017.74-3.342M6.75 15a.75.75 0 100-1.5.75.75 0 000 1.5zm0 0v-3.675A55.378 55.378 0 0112 8.443m-7.007 11.55A5.981 5.981 0 006.75 15.75v-1.5\\\"}]]}]\\n\"])</script><script>self.__next_f.push([1,\"6:D{\\\"name\\\":\\\"AcademicCapIcon\\\"}\\n\"])</script><script>self.__next_f.push([1,\"6:[\\\"$\\\",\\\"svg\\\",null,{\\\"xmlns\\\":\\\"http://www.w3.org/2000/svg\\\",\\\"viewBox\\\":\\\"0 0 24 24\\\",\\\"fill\\\":\\\"currentColor\\\",\\\"aria-hidden\\\":\\\"true\\\",\\\"aria-labelledby\\\":\\\"$undefined\\\",\\\"children\\\":[null,[\\\"$\\\",\\\"path\\\",null,{\\\"d\\\":\\\"M11.7 2.805a.75.75 0 01.6 0A60.65 60.65 0 0122.83 8.72a.75.75 0 01-.231 1.337 49.949 49.949 0 00-9.902 3.912l-.003.002-.34.18a.75.75 0 01-.707 0A50.009 50.009 0 007.5 12.174v-.224c0-.131.067-.248.172-.311a54.614 54.614 0 014.653-2.52.75.75 0 00-.65-1.352 56.129 56.129 0 00-4.78 2.589 1.858 1.858 0 00-.859 1.228 49.803 49.803 0 00-4.634-1.527.75.75 0 01-.231-1.337A60.653 60.653 0 0111.7 2.805z\\\"}],[\\\"$\\\",\\\"path\\\",null,{\\\"d\\\":\\\"M13.06 15.473a48.45 48.45 0 017.666-3.282c.134 1.414.22 2.843.255 4.285a.75.75 0 01-.46.71 47.878 47.878 0 00-8.105 4.342.75.75 0 01-.832 0 47.877 47.877 0 00-8.104-4.342.75.75 0 01-.461-.71c.035-1.442.121-2.87.255-4.286A48.4 48.4 0 016 13.18v1.27a1.5 1.5 0 00-.14 2.508c-.09.38-.222.753-.397 1.11.452.213.901.434 1.346.661a6.729 6.729 0 00.551-1.608 1.5 1.5 0 00.14-2.67v-.645a48.549 48.549 0 013.44 1.668 2.25 2.25 0 002.12 0z\\\"}],[\\\"$\\\",\\\"path\\\",null,{\\\"d\\\":\\\"M4.462 19.462c.42-.419.753-.89 1-1.394.453.213.902.434 1.347.661a6.743 6.743 0 01-1.286 1.794.75.75 0 11-1.06-1.06z\\\"}]]}]\\n\"])</script><script>self.__next_f.push([1,\"3:[\\\"$4\\\",\\\"$5\\\",\\\"$6\\\",[\\\"$\\\",\\\"$L7\\\",null,{}]]\\n8:D{\\\"name\\\":\\\"RootLayout\\\"}\\nb:D{\\\"name\\\":\\\"NotFound\\\"}\\n\"])</script><script>self.__next_f.push([1,\"b:[[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"404: This page could not be found.\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"fontFamily\\\":\\\"system-ui,\\\\\\\"Segoe UI\\\\\\\",Roboto,Helvetica,Arial,sans-serif,\\\\\\\"Apple Color Emoji\\\\\\\",\\\\\\\"Segoe UI Emoji\\\\\\\"\\\",\\\"height\\\":\\\"100vh\\\",\\\"textAlign\\\":\\\"center\\\",\\\"display\\\":\\\"flex\\\",\\\"flexDirection\\\":\\\"column\\\",\\\"alignItems\\\":\\\"center\\\",\\\"justifyContent\\\":\\\"center\\\"},\\\"children\\\":[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"style\\\",null,{\\\"dangerouslySetInnerHTML\\\":{\\\"__html\\\":\\\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\\\"}}],[\\\"$\\\",\\\"h1\\\",null,{\\\"className\\\":\\\"next-error-h1\\\",\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\",\\\"margin\\\":\\\"0 20px 0 0\\\",\\\"padding\\\":\\\"0 23px 0 0\\\",\\\"fontSize\\\":24,\\\"fontWeight\\\":500,\\\"verticalAlign\\\":\\\"top\\\",\\\"lineHeight\\\":\\\"49px\\\"},\\\"children\\\":\\\"404\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\"},\\\"children\\\":[\\\"$\\\",\\\"h2\\\",null,{\\\"style\\\":{\\\"fontSize\\\":14,\\\"fontWeight\\\":400,\\\"lineHeight\\\":\\\"49px\\\",\\\"margin\\\":0},\\\"children\\\":\\\"This page could not be found.\\\"}]}]]}]}]]\\n\"])</script><script>self.__next_f.push([1,\"8:[\\\"$\\\",\\\"html\\\",null,{\\\"lang\\\":\\\"en\\\",\\\"children\\\":[\\\"$\\\",\\\"body\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"$L9\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\"],\\\"loading\\\":\\\"$undefined\\\",\\\"loadingStyles\\\":\\\"$undefined\\\",\\\"loadingScripts\\\":\\\"$undefined\\\",\\\"hasLoading\\\":false,\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$La\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$b\\\",\\\"notFoundStyles\\\":[],\\\"styles\\\":null}]}]}]\\nc:D{\\\"name\\\":\\\"\\\"}\\ne:[]\\n0:[null,[\\\"$\\\",\\\"$L1\\\",n\"])</script><script>self.__next_f.push([1,\"ull,{\\\"buildId\\\":\\\"development\\\",\\\"assetPrefix\\\":\\\"\\\",\\\"initialCanonicalUrl\\\":\\\"/\\\",\\\"initialTree\\\":[\\\"\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{}]},\\\"$undefined\\\",\\\"$undefined\\\",true],\\\"initialSeedData\\\":[\\\"\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{},[[\\\"$L2\\\",\\\"$3\\\"],null]]},[\\\"$8\\\",null]],\\\"initialHead\\\":[false,\\\"$Lc\\\"],\\\"globalErrorComponent\\\":\\\"$d\\\",\\\"missingSlots\\\":\\\"$We\\\"}]]\\n\"])</script><script>self.__next_f.push([1,\"c:[[\\\"$\\\",\\\"meta\\\",\\\"0\\\",{\\\"name\\\":\\\"viewport\\\",\\\"content\\\":\\\"width=device-width, initial-scale=1\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"1\\\",{\\\"charSet\\\":\\\"utf-8\\\"}],[\\\"$\\\",\\\"title\\\",\\\"2\\\",{\\\"children\\\":\\\"Next.js\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"3\\\",{\\\"name\\\":\\\"description\\\",\\\"content\\\":\\\"Generated by Next.js\\\"}]]\\n2:null\\n\"])</script></body></html>"

      49 |
      50 |           // Ensure the icons are rendered
    > 51 |           expect(html).toContain('<svg xmlns="http://www.w3.org/2000/svg"')
         |                        ^
      52 |
      53 |           const modules = [
      54 |             ...logs.matchAll(

      at Object.toContain (development/basic/barrel-optimization/barrel-optimization.test.ts:51:24)

  ● Skipped in Turbopack › optimizePackageImports › should support MUI

    expect(received).toContain(expected) // indexOf

    Expected substring: "button"
    Received string:    ""

      144 |           // Ensure that MUI is working
      145 |           const $ = await next.render$('/mui')
    > 146 |           expect(await $('#button').text()).toContain('button')
          |                                             ^
      147 |           expect(await $('#typography').text()).toContain('typography')
      148 |
      149 |           const browser = await next.browser('/mui')

      at Object.toContain (development/basic/barrel-optimization/barrel-optimization.test.ts:146:45)

  ● Skipped in Turbopack › optimizePackageImports › should support "use client" directive in barrel file

    expect(received).toContain(expected) // indexOf

    Expected substring: "<button>0</button>"
    Received string:    "<!DOCTYPE html><html id=\"__next_error__\"><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"/><link rel=\"preload\" as=\"script\" fetchPriority=\"low\" href=\"/_next/static/chunks/webpack.js\"/><script src=\"/_next/static/chunks/main-app.js\" async=\"\"></script><script src=\"/_next/static/chunks/app-pages-internals.js\" async=\"\"></script><meta name=\"robots\" content=\"noindex\"/><meta name=\"next-error\" content=\"not-found\"/><title>Next.js</title><meta name=\"description\" content=\"Generated by Next.js\"/><script src=\"/_next/static/chunks/polyfills.js\" noModule=\"\"></script></head><body><script src=\"/_next/static/chunks/webpack.js\" async=\"\"></script><script>(self.__next_f=self.__next_f||[]).push([0]);self.__next_f.push([2,null])</script><script>self.__next_f.push([1,\"0:D{\\\"name\\\":\\\"rD\\\"}\\n\"])</script><script>self.__next_f.push([1,\"1:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/app-router.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n4:I[\\\"(app-pages-browser)/__barrel_optimize__?names=Button!=!./node_modules/my-client-lib/index.js\\\",[\\\"app/client-boundary/page\\\",\\\"static/chunks/app/client-boundary/page.js\\\"],\\\"Button\\\"]\\n5:I[\\\"(app-pages-browser)/./node_modules/\"])</script><script>self.__next_f.push([1,\".pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/layout-router.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n6:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/render-from-template-context.js\\\",[\\\"app-pages-intern\"])</script><script>self.__next_f.push([1,\"als\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\na:I[\\\"(app-pages-browser)/./node_modules/.pnpm/file+..+next-repo-23135b0278e30c6789dec604f9bd893732e0dfe416be296bcca64815083d45e5+packages+n_sxayjzg45dwo6cbruaryt6ccq4/node_modules/next/dist/client/components/error-boundary.js\\\",[\\\"app-pages-internals\\\",\\\"static/chunks/app-pages-internals.js\\\"],\\\"\\\"]\\n2:D{\\\"name\\\":\\\"\\\"}\\n3:D{\\\"name\\\":\\\"Page\\\"}\\n3:[\\\"$\\\",\\\"$L4\\\",null,{}]\\n7:D{\\\"name\\\":\\\"RootLayout\\\"}\\n8:D{\\\"name\\\":\\\"NotFound\\\"}\\n\"])</script><script>self.__next_f.push([1,\"8:[[\\\"$\\\",\\\"title\\\",null,{\\\"children\\\":\\\"404: This page could not be found.\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"fontFamily\\\":\\\"system-ui,\\\\\\\"Segoe UI\\\\\\\",Roboto,Helvetica,Arial,sans-serif,\\\\\\\"Apple Color Emoji\\\\\\\",\\\\\\\"Segoe UI Emoji\\\\\\\"\\\",\\\"height\\\":\\\"100vh\\\",\\\"textAlign\\\":\\\"center\\\",\\\"display\\\":\\\"flex\\\",\\\"flexDirection\\\":\\\"column\\\",\\\"alignItems\\\":\\\"center\\\",\\\"justifyContent\\\":\\\"center\\\"},\\\"children\\\":[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[[\\\"$\\\",\\\"style\\\",null,{\\\"dangerouslySetInnerHTML\\\":{\\\"__html\\\":\\\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\\\"}}],[\\\"$\\\",\\\"h1\\\",null,{\\\"className\\\":\\\"next-error-h1\\\",\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\",\\\"margin\\\":\\\"0 20px 0 0\\\",\\\"padding\\\":\\\"0 23px 0 0\\\",\\\"fontSize\\\":24,\\\"fontWeight\\\":500,\\\"verticalAlign\\\":\\\"top\\\",\\\"lineHeight\\\":\\\"49px\\\"},\\\"children\\\":\\\"404\\\"}],[\\\"$\\\",\\\"div\\\",null,{\\\"style\\\":{\\\"display\\\":\\\"inline-block\\\"},\\\"children\\\":[\\\"$\\\",\\\"h2\\\",null,{\\\"style\\\":{\\\"fontSize\\\":14,\\\"fontWeight\\\":400,\\\"lineHeight\\\":\\\"49px\\\",\\\"margin\\\":0},\\\"children\\\":\\\"This page could not be found.\\\"}]}]]}]}]]\\n\"])</script><script>self.__next_f.push([1,\"7:[\\\"$\\\",\\\"html\\\",null,{\\\"lang\\\":\\\"en\\\",\\\"children\\\":[\\\"$\\\",\\\"body\\\",null,{\\\"children\\\":[\\\"$\\\",\\\"$L5\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\"],\\\"loading\\\":\\\"$undefined\\\",\\\"loadingStyles\\\":\\\"$undefined\\\",\\\"loadingScripts\\\":\\\"$undefined\\\",\\\"hasLoading\\\":false,\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L6\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$8\\\",\\\"notFoundStyles\\\":[],\\\"styles\\\":null}]}]}]\\n9:D{\\\"name\\\":\\\"\\\"}\\nb:[]\\n\"])</script><script>self.__next_f.push([1,\"0:[null,[\\\"$\\\",\\\"$L1\\\",null,{\\\"buildId\\\":\\\"development\\\",\\\"assetPrefix\\\":\\\"\\\",\\\"initialCanonicalUrl\\\":\\\"/client-boundary\\\",\\\"initialTree\\\":[\\\"\\\",{\\\"children\\\":[\\\"client-boundary\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{}]}]},\\\"$undefined\\\",\\\"$undefined\\\",true],\\\"initialSeedData\\\":[\\\"\\\",{\\\"children\\\":[\\\"client-boundary\\\",{\\\"children\\\":[\\\"__PAGE__\\\",{},[[\\\"$L2\\\",\\\"$3\\\"],null]]},[\\\"$\\\",\\\"$L5\\\",null,{\\\"parallelRouterKey\\\":\\\"children\\\",\\\"segmentPath\\\":[\\\"children\\\",\\\"client-boundary\\\",\\\"children\\\"],\\\"loading\\\":\\\"$undefined\\\",\\\"loadingStyles\\\":\\\"$undefined\\\",\\\"loadingScripts\\\":\\\"$undefined\\\",\\\"hasLoading\\\":false,\\\"error\\\":\\\"$undefined\\\",\\\"errorStyles\\\":\\\"$undefined\\\",\\\"errorScripts\\\":\\\"$undefined\\\",\\\"template\\\":[\\\"$\\\",\\\"$L6\\\",null,{}],\\\"templateStyles\\\":\\\"$undefined\\\",\\\"templateScripts\\\":\\\"$undefined\\\",\\\"notFound\\\":\\\"$undefined\\\",\\\"notFoundStyles\\\":\\\"$undefined\\\",\\\"styles\\\":null}]]},[\\\"$7\\\",null]],\\\"initialHead\\\":[false,\\\"$L9\\\"],\\\"globalErrorComponent\\\":\\\"$a\\\",\\\"missingSlots\\\":\\\"$Wb\\\"}]]\\n\"])</script><script>self.__next_f.push([1,\"9:[[\\\"$\\\",\\\"meta\\\",\\\"0\\\",{\\\"name\\\":\\\"viewport\\\",\\\"content\\\":\\\"width=device-width, initial-scale=1\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"1\\\",{\\\"charSet\\\":\\\"utf-8\\\"}],[\\\"$\\\",\\\"title\\\",\\\"2\\\",{\\\"children\\\":\\\"Next.js\\\"}],[\\\"$\\\",\\\"meta\\\",\\\"3\\\",{\\\"name\\\":\\\"description\\\",\\\"content\\\":\\\"Generated by Next.js\\\"}]]\\n2:null\\n\"])</script></body></html>"

      166 |         it('should support "use client" directive in barrel file', async () => {
      167 |           const html = await next.render('/client-boundary')
    > 168 |           expect(html).toContain('<button>0</button>')
          |                        ^
      169 |         })
      170 |       }
      171 |     )

      at Object.toContain (development/basic/barrel-optimization/barrel-optimization.test.ts:168:24)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| buildDuration | 11.7s | 11.8s | N/A |
| buildDurationCached | 6.2s | 5s | N/A |
| nodeModulesSize | 196 MB | 196 MB | ⚠️ +716 B |
| nextStartRea..uration (ms) | 428ms | 433ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 29.7 kB | 29.7 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 238 B | 240 B | N/A |
| main-HASH.js gzip | 31.9 kB | 31.9 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 46.9 kB | 46.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 501 B | N/A |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 106 B | 106 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| index.html gzip | 527 B | 527 B | ✓ |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 527 B | 527 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.4 kB | 94.4 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 621 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 47.4 kB | 47.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.94 kB | 1.94 kB | ✓ |
| Overall change | 1.94 kB | 1.94 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/server-components-packages | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 166 kB | 166 kB | ✓ |
| app-page-exp..prod.js gzip | 95.4 kB | 95.4 kB | ✓ |
| app-page-tur..prod.js gzip | 97.2 kB | 97.2 kB | ✓ |
| app-page-tur..prod.js gzip | 91.6 kB | 91.6 kB | ✓ |
| app-page.run...dev.js gzip | 136 kB | 136 kB | ✓ |
| app-page.run..prod.js gzip | 90.2 kB | 90.2 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.6 kB | 14.6 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.7 kB | 9.7 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.43 kB | 9.43 kB | ✓ |
| pages-turbo...prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim...dev.js gzip | 22.7 kB | 22.7 kB | ✓ |
| pages.runtim..prod.js gzip | 22.1 kB | 22.1 kB | ✓ |
| server.runti..prod.js gzip | 49.9 kB | 49.9 kB | ✓ |
| Overall change | 924 kB | 924 kB | ✓ |

</details>

</details>
Commit: 2092cb9adeeb21681e2eda4abcc8bbc91c1367f3

**Review by shuding — COMMENTED** [review]
One future optimization is to have `optOutBundlingPackages` as a Set instead of an array, given that we might do hundreds/thousands of lookups during a compilation.

File: packages/next/src/build/handle-externals.ts [hunk 1]
```diff
@@ -374,28 +359,86 @@ export function makeExternalHandler({
...
+  const shouldBeBundled =
+    isResourceInPackages(
+      resolvedRes,
+      config.transpilePackages,
+      resolvedExternalPackageDirs
+    ) ||
+    (isEsm && isAppLayer) ||
+    (!isAppLayer && config.experimental.bundlePagesExternals)
+
+  const isOptOutBundling = optOutBundlingPackageRegex.test(resolvedRes)
```
Comment (shuding) [inline packages/next/src/build/handle-externals.ts:412]: Might be able to move this into the `if (nodeModulesRegex.test(resolvedRes)) {` below so no unnecessary regex executions.

**Review by shuding — APPROVED** [review]

**Pull request merged**